### ISSUE TITLE
Network module: compact bandwidth formatting

### ISSUE THREAD

**Issue by YodaEmbedding** [Alexays/Waybar#2162]
For the network module, it would be nice to format the bandwidth speeds in a more compact, and also non-varying manner.

```
Current      Desired
420.1MB/s    420 M
420.1kB/s    420 k
420.1B/s     420 B
42.0B/s       42 B
4.2B/s         4 B
```

This shaves off a few unimportant characters for valuable waybar real estate, and prevents the module from varying in width (when wrapped in a monospace font, e.g. `<span font='NotoSansMono-Regular'>{bandwidthDownBytes}</span>`).

**Comment by YodaEmbedding** [comment]
I ended up writing my own custom module:

```python
#!/usr/bin/env python

import subprocess
from time import sleep


def default_interface():
    process = subprocess.run(
        ["ip", "route"], check=True, text=True, capture_output=True
    )
    for line in process.stdout.splitlines():
        if line.startswith("default via"):
            return line.split()[4]
    raise RuntimeError("No default interface found")


def get_rx_tx_bytes(iface):
    with open("/proc/net/dev") as f:
        for line in f:
            if not line.startswith(f"{iface}:"):
                continue
            rx_bytes = int(line.split()[1])
            tx_bytes = int(line.split()[9])
            return rx_bytes, tx_bytes


def format_size(size):
    power_labels = {0: "B", 1: "K", 2: "M", 3: "G", 4: "T"}
    kilo = 2**10
    power = 0
    while size > kilo:
        size /= kilo
        power += 1
    return f"{size:3.0f} {power_labels[power]}"


def main():
    rx_icon = " "
    tx_icon = " "
    num_left = "<span font='NotoSansMono'>"
    num_right = "</span>"
    fmt_str = (
        f"{rx_icon}{num_left}{{rx}}{{unit_suffix}}{num_right}  "
        f"{tx_icon}{num_left}{{tx}}{{unit_suffix}}{num_right}"
    )
    unit_suffix = ""
    iface = default_interface()

    rx_bytes, tx_bytes = get_rx_tx_bytes(iface)
    sleep(1)

    while True:
        prev_rx_bytes, prev_tx_bytes = rx_bytes, tx_bytes
        rx_bytes, tx_bytes = get_rx_tx_bytes(iface)
        drx = format_size(rx_bytes - prev_rx_bytes)
        dtx = format_size(tx_bytes - prev_tx_bytes)
        line = fmt_str.format(rx=drx, tx=dtx, unit_suffix=unit_suffix)
        print(line, flush=True)
        sleep(1)


if __name__ == "__main__":
    main()
```

```json
    "custom/bandwidth": {
        "exec": "$HOME/.config/waybar/scripts/bandwidth"
    },
```

**Comment by John-Dennehy** [comment]
Agreed. Module is currently unusable on vertical bars due to this.  I know I can rotate, but that's just as bad making everything else jump around.

Would also love to see a way to fix it to MB/s etc

**Comment by csskevin** [comment]
Constant width is already possible by using specifications like `<`, `>`, `=`.
e.g. for `bandwidthDownBytes`:
- `{bandwidthDownBytes:<}` aligns the speed and the unit to the left
- `{bandwidthDownBytes:>}` aligns the speed and the unit to the right
- `{bandwidthDownBytes:=}` aligns the speed to the left and the unit to the right

**Comment by jcbevns** [comment]
> I ended up writing my own custom module:
> 
> ```python
> #!/usr/bin/env python
> 
> import subprocess
> from time import sleep
> 
> 
> def default_interface():
>     process = subprocess.run(
>         ["ip", "route"], check=True, text=True, capture_output=True
>     )
>     for line in process.stdout.splitlines():
>         if line.startswith("default via"):
>             return line.split()[4]
>     raise RuntimeError("No default interface found")
> 
> 
> def get_rx_tx_bytes(iface):
>     with open("/proc/net/dev") as f:
>         for line in f:
>             if not line.startswith(f"{iface}:"):
>                 continue
>             rx_bytes = int(line.split()[1])
>             tx_bytes = int(line.split()[9])
>             return rx_bytes, tx_bytes
>     raise RuntimeError("Interface not found")
> 
> 
> def format_size(size):
>     power_labels = {0: "B", 1: "K", 2: "M", 3: "G", 4: "T"}
>     kilo = 2**10
>     power = 0
>     while size > kilo:
>         size /= kilo
>         power += 1
>     return f"{size:3.0f} {power_labels[power]}"
> 
> 
> def main():
>     refresh_interval = 2
>     rx_icon = " "
>     tx_icon = " "
>     num_left = "<span font='NotoSansMono'>"
>     num_right = "</span>"
>     fmt_str = (
>         f"{rx_icon}{num_left}{{rx}}{{unit_suffix}}{num_right}  "
>         f"{tx_icon}{num_left}{{tx}}{{unit_suffix}}{num_right}"
>     )
>     unit_suffix = ""
>     iface = default_interface()
> 
>     rx_bytes, tx_bytes = get_rx_tx_bytes(iface)
> 
>     while True:
>         prev_rx_bytes, prev_tx_bytes = rx_bytes, tx_bytes
>         rx_bytes, tx_bytes = get_rx_tx_bytes(iface)
>         drx = format_size((rx_bytes - prev_rx_bytes) / refresh_interval)
>         dtx = format_size((tx_bytes - prev_tx_bytes) / refresh_interval)
>         line = fmt_str.format(rx=drx, tx=dtx, unit_suffix=unit_suffix)
>         print(line, flush=True)
>         sleep(refresh_interval)
> 
> 
> if __name__ == "__main__":
>     main()
> ```
> 
> config:
> 
> ```json
>     "custom/bandwidth": {
>         "exec": "$HOME/.config/waybar/scripts/bandwidth"
>     },
> ```

Does this solve the variable width issue? @YodaEmbedding ?

**Comment by Ctmax-ui** [comment]
Yes we need an way to configure between mb, kb, b

**Comment by layus** [comment]
Having implemented the <, > and = format specifiers, your format is definitely doable. But how would you deal with overflows and underflows ? Would that alter the fixed width ? I guess you could be happy with truncating values, but that seems odd to me.

**Comment by layus** [comment]
Or even better, define your own format modifiers and implement that in https://github.com/Alexays/Waybar/blob/master/include/util/format.hpp.

One could have {bandwidthDownBytes:M} to display it in megabytes. Then {bandwidthDownBytes:M>8} would format it as a fixed size of 8 characters. For that specific format I would also not include the units by default in the output, allowing users to display them as they see fit.

There are many corner cases to thing about, like decimals if you want three significant numbers with a network that goes up to 10MB/s.

But overall, I think passing a unit should turn the pow_format number into a normal float number.
